### ISSUE TITLE
BIP322 verification

### ISSUE THREAD

**Issue by mrv777** [bitcoinjs/bitcoinjs-lib#1918]
With BIP-322 starting to be used more, I feel everything is in this library to verify a BIP-322 signature but can't seem to figure out the right pieces to put together.  Given an `address`, `signature`, and the `message` we should be able to verify the given address signed the given message.

base64.decode(signature) and bitcoin.address.toOutputScript(address) seems to be a piece, but the rest I'm not sure about .

There are some tests in the documentation:
https://bips.xyz/322#message-signing

Edit: My current thinking is to create an unsigned PSBT that follows the structure here: https://bips.xyz/322#full and take `base64.decode(signature)` and inject it somehow into that, then you can check if its still valid and if so assume that signature is valid for a bip-322 message from that account with that message.

This is what I'm doing to create the unsigned part:
```
function createToSpendTx(address: string, message: string, network: any) {
  const { prevoutHash, prevoutIndex, sequence } = bip322TransactionToSignValues;

  const script = bitcoin.address.toOutputScript(
    address,
    bitcoin.networks.bitcoin
  ); //message_challenge 

  const hash = bip0322Hash(message);
  const commands = [0, Buffer.from(hash)]; // OP_0 PUSH32[ message_hash ]?
  const scriptSig = bitcoin.script.compile(commands);

  const virtualToSpend = new bitcoin.Transaction();
  virtualToSpend.version = 0;
  virtualToSpend.addInput(Buffer.from(prevoutHash), prevoutIndex, sequence, scriptSig);
  virtualToSpend.addOutput(script, 0);
  return { virtualToSpend, script };
}
function createToSignTx(toSpendTxHex: Buffer, script: Buffer, network: any) {
  const virtualToSign = new bitcoin.Psbt({ network: network });
  virtualToSign.setVersion(0);
  const prevTxHash = toSpendTxHex;
  const prevOutIndex = 0;
  const toSignScriptSig = bitcoin.script.compile([106]); //OP Code 106 for OP_RETURN

  virtualToSign.addInput({
    hash: prevTxHash,
    index: prevOutIndex,
    sequence: 0,
    witnessUtxo: { script, value: 0 }
  });

  virtualToSign.addOutput({ script: toSignScriptSig, value: 0 });
  return virtualToSign;
}
```

Edit 2: What I was thinking would be next then:
```
const signature = base64.decode(signature);
const decodedWitness = signature.slice(2, 66); //First two elements seem to be 1(?) and then 64(length)

virtualToSign.updateInput(0, {
      tapKeySig: serializeTaprootSignature(Buffer.from(decodedWitness))
  });
    virtualToSign.finalizeAllInputs();
    const toSignTx = virtualToSign.extractTransaction();
```

This creates a transaction with no errors, but not sure how to validate it because if I change the address or message variables so it should be invalid, it should throw an error

Thank for any insights

**Comment by junderw** [comment]
Psbt is not designed for anything except for creating valid spending transactions.

You should use the code inside Psbt to help you write signers and verifiers for the various scripts.

For p2wpkh it might be something like this: (it works, but it's hastily slapped together. Feel free to modify it to fit your needs)

```js
const assert = require('assert');
const bitcoinJsLib = require('bitcoinjs-lib');
const tinysecp = require('tiny-secp256k1');
bitcoinJsLib.initEccLib(tinysecp);
const { ECPairFactory } = require('ecpair');
const ecpair = ECPairFactory(tinysecp);

/**
 * Simple verification for BIP322 signatures
 * @param {string} message A message that is signed ie "Hello world"
 * @param {string} address The address that signed it.
 * @param {string} signature The signature (which is an encoded transaction as per BIP322)
 * @returns {boolean}
 */
function verifyBip322(message, address, signature) {
  const output = bitcoinJsLib.payments.p2wpkh({ address }).output;
  if (output === undefined) {
    throw new Error('Only supports p2wpkh addresses for now');
  }
  const hash = taggedHash(
    'BIP0322-signed-message',
    Buffer.from(message, 'utf8'),
  );
  const sigBuff = Buffer.from(signature, 'base64');

  const toSpend = createToSpend(hash, output);
  const toSign = createToSign(toSpend, sigBuff);

  return validateP2wpkh(toSign);
}

/**
 * Signs the message according to BIP322 for P2WPKH
 * @param {string} message
 * @param {Buffer} privateKey
 * @returns {string}
 */
function signBip322(message, privateKey) {
  const hash = taggedHash(
    'BIP0322-signed-message',
    Buffer.from(message, 'utf8'),
  );
  const keyPair = ecpair.fromPrivateKey(privateKey);
  const output = bitcoinJsLib.payments.p2wpkh({
    pubkey: keyPair.publicKey,
  }).output;
  const toSpend = createToSpend(hash, output);
  const toSign = createToSign(toSpend);
  return signP2wpkh(toSign, keyPair);
}

/**
 * Tagged hashing
 * @param {string} prefix
 * @param {Buffer} data
 * @returns {Buffer}
 */
function taggedHash(prefix, data) {
  const prefixHash = bitcoinJsLib.crypto.sha256(prefix);
  return bitcoinJsLib.crypto.sha256(
    Buffer.concat([prefixHash, prefixHash, data]),
  );
}

/**
 * Creates the toSpend tx
 * @param {Buffer} messageHash
 * @param {Buffer} outputScript
 * @returns {bitcoinJsLib.Transaction}
 */
function createToSpend(messageHash, outputScript) {
  const { OP_0 } = bitcoinJsLib.opcodes;
  const toSpend = new bitcoinJsLib.Transaction();
  toSpend.version = 0;
  toSpend.locktime = 0;
  toSpend.addInput(
    Buffer.alloc(32, 0),
    0xffffffff,
    0,
    Buffer.concat([Buffer.from([OP_0, 0x20]), messageHash]),
  );
  toSpend.addOutput(outputScript, 0);

  return toSpend;
}

/**
 * Creates the toSign transaction
 * @param {bitcoinJsLib.Transaction} toSpend
 * @param {Buffer?} signature
 * @returns {bitcoinJsLib.Transaction}
 */
function createToSign(toSpend, signature = Buffer.from([])) {
  const { OP_RETURN } = bitcoinJsLib.opcodes;
  if (signature.length === 65) {
    throw new Error('Legacy signatures can be handled with bitcoinjs-message');
  }
  let toSign;
  let isTx;
  try {
    toSign = bitcoinJsLib.Transaction.fromBuffer(signature);
    isTx = true;
  } catch (e) {
    // If serialization fails, it is a witnessStack only (simplified signature)
    toSign = new bitcoinJsLib.Transaction();
    isTx = false;
  }

  if (isTx) {
    assert(toSign.ins.length === 1);
    assert(toSign.outs.length === 1);
    assert(toSign.ins[0].index === 0);
    assert(toSign.outs[0].value === 0);
    assert(toSign.outs[0].script.length === 1);
    assert(toSign.outs[0].script[0] === OP_RETURN);
    const toSpendHash = toSpend.getHash();
    const input = toSign.ins[0];
    assert.strictEqual(input.hash, toSpendHash);
  } else {
    toSign.version = 0;
    toSign.locktime = 0;
    toSign.addOutput(Buffer.from([OP_RETURN]), 0);
    toSign.addInput(toSpend.getHash(), 0, 0);
    if (signature.length > 0) {
      // We assume 1 input, so witness stack will only have one stack.
      // Skip the item count (first byte)
      toSign.ins[0].witness = bitcoinJsLib.script.toStack(signature.slice(1));
    }
  }

  return toSign;
}

/**
 * Validates the toSign transaction
 * @param {bitcoinJsLib.Transaction} toSign
 * @returns {boolean}
 */
function validateP2wpkh(toSign) {
  // Verify the signatures in toSign transaction
  const [witnessSignature, pubkey] = toSign.ins[0].witness;
  // p2wpkh uses p2pkh script for creating the sighash for some reason
  const signingOutput = bitcoinJsLib.payments.p2pkh({ pubkey }).output;
  const hashForSigning = toSign.hashForWitnessV0(
    0,
    signingOutput,
    0,
    bitcoinJsLib.Transaction.SIGHASH_ALL,
  );

  const sig = bitcoinJsLib.script.signature.decode(witnessSignature);
  assert(sig.hashType === bitcoinJsLib.Transaction.SIGHASH_ALL);

  const pair = ecpair.fromPublicKey(pubkey);

  return pair.verify(hashForSigning, sig.signature);
}

/**
 * Signs the message with P2WPKH
 * @param {bitcoinJsLib.Transaction} toSign
 * @param {import('ecpair').ECPairInterface} keypair
 * @returns {string}
 */
function signP2wpkh(toSign, keypair) {
  const signingOutput = bitcoinJsLib.payments.p2pkh({
    pubkey: keypair.publicKey,
  }).output;
  const hashForSigning = toSign.hashForWitnessV0(
    0,
    signingOutput,
    0,
    bitcoinJsLib.Transaction.SIGHASH_ALL,
  );
  const signature = keypair.sign(hashForSigning);
  const encodedSignature = bitcoinJsLib.script.signature.encode(
    signature,
    bitcoinJsLib.Transaction.SIGHASH_ALL,
  );
  const compiled = Buffer.concat([
    Buffer.from([0x02]),
    bitcoinJsLib.script.compile([encodedSignature, keypair.publicKey]),
  ]);
  return compiled.toString('base64');
}

/**
 * This function is for testing the above functions
 */
function test() {
  const testSignature =
    'AkcwRAIgZRfIY3p7/DoVTty6YZbWS71bc5Vct9p9Fia83eRmw2QCICK/ENGfwLtptFluMGs2KsqoNSk89pO7F29zJLUx9a/sASECx/EgAxlkQpQ9hYjgGu6EBCPMVPwVIVJqO4XCsMvViHI=';
  const testMessage = 'Hello World';
  const testAddress = 'bc1q9vza2e8x573nczrlzms0wvx3gsqjx7vavgkx0l';
  // The sign function only accepts private key buffers since WIF format is dead
  // It is only really used for displaying test cases for private keys, no wallets support them by default
  const testPrivateKey = ecpair.fromWIF(
    'L3VFeEujGtevx9w18HD1fhRbCH67Az2dpCymeRE1SoPK6XQtaN2k',
  ).privateKey;

  // Should output
  // true
  // true
  // false
  const output1 = verifyBip322(testMessage, testAddress, testSignature);
  const output2 = verifyBip322(
    testMessage,
    testAddress,
    signBip322(testMessage, testPrivateKey),
  );
  const output3 = verifyBip322(
    'Not hello world',
    testAddress,
    signBip322(testMessage, testPrivateKey),
  );
  console.log('Verify BIP test case   :', output1);
  console.log('Verify our own signing :', output2);
  console.log('Verify mismatch (false):', output3);
  assert(output1);
  assert(output2);
  assert(output3 === false);
}

test();
```

**Comment by mrv777** [comment]
This code looks great, thanks @junderw!
Looks like now I have to look into modifying this for p2tr (bc1p...) addresses, as that is what I am trying to use it with, or figuring out how to not use the taproot address for this

**Comment by junderw** [comment]
If you come up with something feel free to post it here.

The only parts you should need to change are the address parsing at the beginning of the verify function, and then the two *P2wpkh functions.

**Comment by mrv777** [comment]
Thanks!
So far it looks like changing to p2tr causes toSign.ins[0].witness to only have one element in the array so the pubKey is missing here:

`const [witnessSignature, pubkey] = toSign.ins[0].witness;`

and the function `const signingOutput = bitcoinJsLib.payments.p2tr({ pubkey }).output;` fails with the error:

TypeError: Not enough data

I will keep exploring this, the insights are most welcome~

**Comment by junderw** [comment]
> So far it looks like changing to p2tr causes toSign.ins[0].witness to only have one element in the array so the pubKey is missing here:

If you don't have the slightest understanding of how p2tr works, I suggest reading the related BIPs (341, 342) to learn how it works, how transaction validation works etc.

tl;dr, the pubkey is in the actual output script (and encoded directly into the p2tr address)

**Comment by junderw** [comment]
This is probably filled with bugs, but I added a tweak flag to signing in order to perform the tweaking recommended in BIP 341 and required in BIP86.

It only can verify taproot key spends without any annex included. If you don't know what that means, it might not be a good idea to use this in production.

That said, it is probably a good starting point.

```js
const assert = require('assert');
const bitcoinJsLib = require('bitcoinjs-lib');
const tinysecp = require('tiny-secp256k1');
bitcoinJsLib.initEccLib(tinysecp);
const { ECPairFactory } = require('ecpair');
const ecpair = ECPairFactory(tinysecp);

/**
 * Simple verification for BIP322 signatures
 * @param {string} message A message that is signed ie "Hello world"
 * @param {string} address The address that signed it.
 * @param {string} signature The signature (which is an encoded transaction as per BIP322)
 * @returns {boolean}
 */
function verifyBip322(message, address, signature) {
  const { output, pubkey } = bitcoinJsLib.payments.p2tr({ address });
  if (output === undefined) {
    throw new Error('Only supports p2tr addresses');
  }
  const hash = taggedHash(
    'BIP0322-signed-message',
    Buffer.from(message, 'utf8'),
  );
  const sigBuff = Buffer.from(signature, 'base64');

  const toSpend = createToSpend(hash, output);
  const toSign = createToSign(toSpend, sigBuff);

  return validateP2tr(toSign, pubkey);
}

/**
 * Signs the message according to BIP322 for P2WPKH
 * @param {string} message
 * @param {Buffer} privateKey
 * @param {boolean} tweak
 * @returns {[string, string]} address for validation and signature
 */
function signBip322(message, privateKey, tweak = false) {
  const hash = taggedHash(
    'BIP0322-signed-message',
    Buffer.from(message, 'utf8'),
  );
  let keyPair = ecpair.fromPrivateKey(privateKey);
  let pubkey = keyPair.publicKey.subarray(1);
  if (tweak) {
    const tweakHash = bitcoinJsLib.crypto.taggedHash('TapTweak', pubkey);
    keyPair = keyPair.tweak(tweakHash);
    pubkey = keyPair.publicKey.subarray(1);
  }
  const { output, address } = bitcoinJsLib.payments.p2tr({
    pubkey,
  });
  const toSpend = createToSpend(hash, output);
  const toSign = createToSign(toSpend);
  return [address, signP2tr(toSign, keyPair)];
}

/**
 * Tagged hashing
 * @param {string} prefix
 * @param {Buffer} data
 * @returns {Buffer}
 */
function taggedHash(prefix, data) {
  const prefixHash = bitcoinJsLib.crypto.sha256(prefix);
  return bitcoinJsLib.crypto.sha256(
    Buffer.concat([prefixHash, prefixHash, data]),
  );
}

/**
 * Creates the toSpend tx
 * @param {Buffer} messageHash
 * @param {Buffer} outputScript
 * @returns {bitcoinJsLib.Transaction}
 */
function createToSpend(messageHash, outputScript) {
  const { OP_0 } = bitcoinJsLib.opcodes;
  const toSpend = new bitcoinJsLib.Transaction();
  toSpend.version = 0;
  toSpend.locktime = 0;
  toSpend.addInput(
    Buffer.alloc(32, 0),
    0xffffffff,
    0,
    Buffer.concat([Buffer.from([OP_0, 0x20]), messageHash]),
  );
  toSpend.addOutput(outputScript, 0);

  return toSpend;
}

/**
 * Creates the toSign transaction
 * @param {bitcoinJsLib.Transaction} toSpend
 * @param {Buffer?} signature
 * @returns {bitcoinJsLib.Transaction}
 */
function createToSign(toSpend, signature = Buffer.from([])) {
  const { OP_RETURN } = bitcoinJsLib.opcodes;
  if (signature.length === 65) {
    throw new Error('Legacy signatures can be handled with bitcoinjs-message');
  }
  let toSign;
  let isTx;
  try {
    toSign = bitcoinJsLib.Transaction.fromBuffer(signature);
    isTx = true;
  } catch (e) {
    // If serialization fails, it is a witnessStack only (simplified signature)
    toSign = new bitcoinJsLib.Transaction();
    isTx = false;
  }

  if (isTx) {
    assert(toSign.ins.length === 1);
    assert(toSign.outs.length === 1);
    assert(toSign.ins[0].index === 0);
    assert(toSign.outs[0].value === 0);
    assert(toSign.outs[0].script.length === 1);
    assert(toSign.outs[0].script[0] === OP_RETURN);
    const toSpendHash = toSpend.getHash();
    const input = toSign.ins[0];
    assert.strictEqual(input.hash, toSpendHash);
  } else {
    toSign.version = 0;
    toSign.locktime = 0;
    toSign.addOutput(Buffer.from([OP_RETURN]), 0);
    toSign.addInput(toSpend.getHash(), 0, 0);
    if (signature.length > 0) {
      // We assume 1 input, so witness stack will only have one stack.
      // Skip the item count (first byte)
      toSign.ins[0].witness = bitcoinJsLib.script.toStack(
        signature.subarray(1),
      );
    }
  }

  return toSign;
}

/**
 * Validates the toSign transaction
 * @param {bitcoinJsLib.Transaction} toSign
 * @param {Buffer} pubkey the 32 byte xonly pubkey
 * @returns {boolean}
 */
function validateP2tr(toSign, pubkey) {
  // Verify the signatures in toSign transaction
  const [schnorrSig] = toSign.ins[0].witness;
  // p2wpkh uses p2pkh script for creating the sighash for some reason
  const hashForSigning = toSign.hashForWitnessV1(
    0,
    [bitcoinJsLib.payments.p2tr({ pubkey }).output],
    [0],
    bitcoinJsLib.Transaction.SIGHASH_DEFAULT,
  );

  // SIGHASH_DEFAULT has no sighash byte
  assert(schnorrSig.length === 64);

  const pair = ecpair.fromPublicKey(Buffer.concat([Buffer.from([2]), pubkey]));

  return pair.verifySchnorr(hashForSigning, schnorrSig);
}

/**
 * Signs the message with P2WPKH
 * @param {bitcoinJsLib.Transaction} toSign
 * @param {import('ecpair').ECPairInterface} keypair
 * @returns {string}
 */
function signP2tr(toSign, keypair) {
  const signingOutput = bitcoinJsLib.payments.p2tr({
    pubkey: keypair.publicKey.subarray(1),
  }).output;
  const hashForSigning = toSign.hashForWitnessV1(
    0,
    [signingOutput],
    [0],
    bitcoinJsLib.Transaction.SIGHASH_DEFAULT,
  );
  const signature = keypair.signSchnorr(hashForSigning);
  const compiled = Buffer.concat([
    Buffer.from([0x01]),
    bitcoinJsLib.script.compile([signature]),
  ]);
  return compiled.toString('base64');
}

/**
 * This function is for testing the above functions
 */
function test() {
  const testSignature =
    'AUAOSzupT68RMtE9FQMv4LQ0secNMzbIXvsNyYa7h/+oDu6TeyyJtnWu6IluMj/AxFdoEek7CXNFsHup/HEqSntk';
  const testMessage = 'Hello World';
  const testAddress =
    'bc1pclcjqqcev3pfg0v93rsp4m5yqs3uc48uz5s4y63mshptpj743peq6x880p';
  // The sign function only accepts private key buffers since WIF format is dead
  // It is only really used for displaying test cases for private keys, no wallets support them by default
  const testPrivateKey = ecpair.fromWIF(
    'L3VFeEujGtevx9w18HD1fhRbCH67Az2dpCymeRE1SoPK6XQtaN2k',
  ).privateKey;

  const output1 = verifyBip322(testMessage, testAddress, testSignature);
  const output2 = verifyBip322(
    testMessage,
    testAddress,
    signBip322(testMessage, testPrivateKey),
  );
  const output3 = verifyBip322(
    'Not hello world',
    testAddress,
    signBip322(testMessage, testPrivateKey),
  );
  const output4 = verifyBip322(
    testMessage,
    testAddress,
    signBip322(testMessage, testPrivateKey, true),
  );
  console.log('Verify BIP test case   :', output1);
  console.log('Verify our own signing :', output2);
  console.log('Verify mismatch (false):', output3);
  console.log('Verify mismatch (false):', output4);
  assert(output1);
  assert(output2);
  assert(output3 === false);
  assert(output4 === false);
}

test();
```

**Comment by robbierogers** [comment]
@junderw I've tried to run both of these examples and they both fail the first assertion. Is it likely to be something wrong with my environment?

```
Verify BIP test case   : false
Verify our own signing : true
Verify mismatch (false): false
```

```
Verify BIP test case   : false
Verify our own signing : true
Verify mismatch (false): false
Verify mismatch (false): false
```

```
"dependencies": {
    "assert": "^2.0.0",
    "bitcoinjs-lib": "^6.1.1",
    "ecpair": "^2.1.0",
    "tiny-secp256k1": "^2.2.1"
  }
```

Node v18.16.0

**Comment by junderw** [comment]
It worked when I ran it.

If you find out the problem, post a reply with the fix.

Thanks!

**Comment by thienma1258** [comment]
@robbierogers have you fixed it yet I have the same problem as you.

**Comment by mrv777** [comment]
> @junderw I've tried to run both of these examples and they both fail the first assertion. Is it likely to be something wrong with my environment?
> 
> ```
> Verify BIP test case   : false
> Verify our own signing : true
> Verify mismatch (false): false
> ```
> 
> ```
> Verify BIP test case   : false
> Verify our own signing : true
> Verify mismatch (false): false
> Verify mismatch (false): false
> ```
> 
> ```
> "dependencies": {
>     "assert": "^2.0.0",
>     "bitcoinjs-lib": "^6.1.1",
>     "ecpair": "^2.1.0",
>     "tiny-secp256k1": "^2.2.1"
>   }
> ```
> 
> Node v18.16.0

Not sure if it helps, but I didn't use any `assert` package.  I just used the internal node.js module.
Are there any other errors you are getting to see exactly what check its failing on?

**Comment by junderw** [comment]
Perhaps I copy pasted it wrong.

I double checked and it should work. Would be better if someone can test it on a known good signature that exists somewhere.

**Comment by mrv777** [comment]
Yeah, it works great for me.  I modified it some for my use, but the example above worked for me and helped a ton.  Thank again!

**Comment by peopledrivemecrazy** [comment]
I am trying to use Hiro Wallet's BIP322 signing, and I can get the signature from the wallet extension. Now when I tried using the above example, the verification failed. Did I implement it wrong?

the code is here https://github.com/peopledrivemecrazy/hiro-wallet-sveltekit/blob/master/src/routes/api/%2Bserver.ts#L14

**Comment by junderw** [comment]
We recently released a change in hash algorithm dependency.

The old dependency treated hashing a string as equivalent to Buffer.from(string, 'utf8') but the new one treats it differently.

That was why the code I wrote worked back when I first wrote it but not now. Thanks @peopledrivemecrazy 

I fixed the code in my comment

**Comment by peopledrivemecrazy** [comment]
Thank you @junderw

**Comment by timechainz** [comment]
@junderw 

In signBip322() should the address be derived using internalPubKey?  The signBip322 works insomuch as it produces a signature that I can verify with the verifyBip322 function but the address it generates is weird.  

Here's the address I get in Sparrow


<img width="815" alt="Screenshot 2023-07-11 at 11 54 46 PM" src="https://github.com/bitcoinjs/bitcoinjs-lib/assets/126361786/efa7cdb5-ef0d-463a-9535-82dbb5a50e1a">

whereas signBip322() produces bc1ptcc6qg0y4hs07l2k4szsglayhqlkgdr8f25x5cunek003p6j7wfqjuwtpg


``` 
const ecc = require("tiny-secp256k1");
const {mnemonicToSeedSync, generateMnemonic} = require("bip39");
const {initEccLib, payments} = require("bitcoinjs-lib");
const { BIP32Factory } = require('bip32');
const bip32 = BIP32Factory(ecc);

initEccLib(ecc);

const mnemonic = 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about';
const passphrase = 'abandon'
const path = "m/86'/0'/0'/0/2"
const expected = 'bc1py874z3vfveuvsqjykdfnuuqymmmclr4gswq0hv4rsmz6trgcttrspyrggj'

const seed = mnemonicToSeedSync(mnemonic, passphrase)
const node = bip32.fromSeed(seed)
const childNode = node.derivePath(path)

const address1 = payments.p2tr({ internalPubkey: childNode.publicKey.subarray(1)}).address
console.log('address using internalPubKey: ' + address1)

const address2 = payments.p2tr({ pubkey: childNode.publicKey.subarray(1)}).address
console.log('address using pubkey: ' + address2)
``` 
```
address using internalPubKey: bc1py874z3vfveuvsqjykdfnuuqymmmclr4gswq0hv4rsmz6trgcttrspyrggj
address using pubkey: bc1ptcc6qg0y4hs07l2k4szsglayhqlkgdr8f25x5cunek003p6j7wfqjuwtpg

```

**Comment by junderw** [comment]
Take another look.

That's exactly what the `tweak` boolean flag is doing.

**Comment by timechainz** [comment]
Aaaaaaaaahhhhhhhhh, of course!  It works perfectly with tweak set to true.  Really appreciate the help.

**Comment by junderw** [comment]
As I said in my comment before:

***Please don't use this in any app that deals with real money or things that have real monetary value...***

It should be a starting point for someone to create a more robust library.

If you didn't even see the tweak argument nor understood what it meant........ then I hope you're just playing around and not trying to build something that relies on these signatures for actual monetary value.

**Comment by junderw** [comment]
I have hidden the code sections behind a clickable dropdown to discourage people searching from finding it without actually reading the thread.

Also, to make sure that people copy pasting understand that this code should be assumed to be bug filled, and only used after full verification of the code.

**Comment by winktool** [comment]
> Aaaaaaaaahhhhhhhhh, of course! It works perfectly with tweak set to true. Really appreciate the help.

plz how to set? which param?
